features/bsm/daemon/src/main/java/org/opennms/netmgt/bsm/daemon/Bsmd.java

```diff
@@ -196,7 +203,70 @@ public void serviceInterfaceOrNodeDeleted(Event e) {
+                    }
+                }
+
+                affectedBusinessServices.addAll(applicationIdToBusinessService.get(e.getParm("applicationId").getValue().getContent()));
+            } else {
+                final Map<String, Set<BusinessService>> reductionKeyToBusinessService = new TreeMap<>();
+
+                for (final BusinessService businessService : businessServices) {
+                    // map ip services edge keys to business services
+                    for(final IpServiceEdge edge : businessService.getIpServiceEdges()) {
+                        for (final String reductionKey : edge.getReductionKeys()) {
+                            if (!reductionKeyToBusinessService.containsKey(reductionKey)) {
+                                reductionKeyToBusinessService.put(reductionKey, new HashSet<>());
+                            }
+                            reductionKeyToBusinessService.get(reductionKey).add(businessService);
+                        }
+                    }
+                    // map reduction key edge keys to business services
+                    for(final ReductionKeyEdge edge : businessService.getReductionKeyEdges()) {
+                        for (final String reductionKey : edge.getReductionKeys()) {
+                            if (!reductionKeyToBusinessService.containsKey(reductionKey)) {
+                                reductionKeyToBusinessService.put(reductionKey, new HashSet<>());
+                            }
+                            reductionKeyToBusinessService.get(reductionKey).add(businessService);
+                        }
+                    }
+                }
+
+                if (EventConstants.NODE_DELETED_EVENT_UEI.equals(e.getUei())) {
+                    affectedBusinessServices.addAll(reductionKeyToBusinessService.get(String.format("uei.opennms.org/nodes/nodeDown::%d", e.getNodeid())));
+                }
+
+                if (EventConstants.INTERFACE_DELETED_EVENT_UEI.equals(e.getUei())) {
+                    affectedBusinessServices.addAll(reductionKeyToBusinessService.get(String.format("uei.opennms.org/nodes/interfaceDown::%d:%s", e.getNodeid(), e.getInterface())));
+                }
+
+                if (EventConstants.SERVICE_DELETED_EVENT_UEI.equals(e.getUei())) {
+                    affectedBusinessServices.addAll(reductionKeyToBusinessService.get(String.format("uei.opennms.org/nodes/nodeLostService::%d:%s:%s", e.getNodeid(), e.getInterface(), e.getService())));
+                }
+            }
+
             reloadConfigurationAt = System.currentTimeMillis() + RELOAD_DELAY;
+
+            final EventBuilder eventBuilder = new EventBuilder(EventConstants.BUSINESS_SERVICE_GRAPH_INVALIDATED, "bsmd");
+            eventBuilder.addParam("businessServiceNames", affectedBusinessServices.stream().map(b -> String.valueOf(b.getName())).sorted().collect(Collectors.toList()));
```

Comment:
Should we create one alarm per business service instead?

Reply:
What do you think makes more sense? I don't care about that. I can also generate an event for each affected business service... what do you think?

Reply:
What led me here was looking at the reduction key for the alarm associated with the event, seems like it would make more sense to de-duplicate these events into alarms based on individual business services, as opposed to sets of business services.

Reply:
Ok

Reply:
I think per BSM would be better, since you can use the alarms as some kind of todo list and process/fix each step by step. Also, you could have more teams/coworkers that are responsible for different BSM. The alarm acknowledging logic fits better then.

Reply:
@christianpape Will test this PR asap.

Reply:
done